src/main/java/io/nats/client/impl/MessageQueue.java

```diff
@@ -27,39 +28,30 @@
     private final static int RUNNING = 1;
     private final static int DRAINING = 2;
 
-    public static final int MAX_SPINS = 200; // Default max spins
-    public static final int SPIN_WAIT = 50;
-    public static final int MAX_SPIN_TIME = SPIN_WAIT * MAX_SPINS;
-
     private final AtomicLong length;
     private final AtomicLong sizeInBytes;
     private final AtomicInteger running;
     private final boolean singleThreadedReader;
-    private final int maxSpins;
-    private final ConcurrentLinkedQueue<NatsMessage> queue;
-    private final ConcurrentLinkedQueue<Thread> waiters;
+    private final LinkedBlockingQueue<NatsMessage> queue;
     private final Lock filterLock;
+    private final NatsMessage poisonPill;
 
-    MessageQueue(boolean singleReaderMode) {
-        this.queue = new ConcurrentLinkedQueue<>();
+    MessageQueue(boolean singleReaderMode, int publishHighwaterMark) {
+        this.queue = publishHighwaterMark > 0 ? new LinkedBlockingQueue<NatsMessage>(publishHighwaterMark) : new LinkedBlockingQueue<NatsMessage>();
         this.running = new AtomicInteger(RUNNING);
         this.sizeInBytes = new AtomicLong(0);
         this.length = new AtomicLong(0);
 
+        // The poisonPill is used to stop poll and accumulate when the queue is stopped
+        this.poisonPill = new NatsMessage("_poison", null, NatsConnection.EMPTY_BODY, false);
+
         this.filterLock = new ReentrantLock();
         
-        this.waiters = new ConcurrentLinkedQueue<>();
         this.singleThreadedReader = singleReaderMode;
+    }
 
-        String os = System.getProperty("os.name");
-        os = os != null ? os.toLowerCase() : "";
-
-        // Windows does not like spin locks, see issue #224
-        if(os.contains("windows")) {
-            this.maxSpins = 0;

```

Comment:
Interested in how this affects windows perf. MC was running benchmark on windows, fyi.

Reply:
no more spinning so this is no longer needed. windows looks ok with the blocking queue

Reply:
I mean to understand why MC was seeing a much different behavior then when you ran against demo. I assume you were not on Windows.

Reply:
i think windows was slower to lock or maybe had lower resolution on the locksupport timeout. So things just "took longer" when I tried to spin wait.

Reply:
If possible we should test and see if we see same behavior I saw at MC.

Reply:
did anyone take notes?

Reply:
I can give you the info.